File: LeetCode/Apple/TopQuestions/Trees_and_Graphs/maxDepth.cpp
//how to define a tree in Cpp
#include <iostream>
#include <map>
#include <vector>

using namespace std;
struct TreeNode{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode (int x):val(x),left(NULL),right(Null){}
};

class Solution{
public:
  int maxDepth(TreeNode * root)
  {
    if (root == NULL)
    {
      return 0;
    }
    else
    {
      int left_height = maxDepth(root->left);
      int right_height = maxDepth(root->right);
      return max(left_height,right_height)+1;
    }
  }
};

int main()
{
  TreeNode node *new node = {3,9,20,NULL,NULL,15,7};
  Solution output;
  cout << "The max depth is:" << output.maxDepth(node)<< endl;
  return 0;
}

File: LeetCode/Apple/TopQuestions/Trees_and_Graphs/maxDepth.py

'''
Given binary tree [3,9,20,null,null,15,7]

  3
   / \
  9  20
    /  \
   15   7
'''
#Definition for a binary tree node.
class TreeNode:
    def __init__(self,x):
        self.val = x
        self.left = None
        self.right = None

#Depth FIrst search
#Recursion
class Solution:
    def maxDepth(self,root:TreeNode) -> int:
        if root is None:
            return 0
        else:
            left_height = self.maxDepth(root.left)
            right_height = self.maxDepth(root.right)
            return max(left_height,right_height)+1

if __name__ == "__main__":
    root = [3,9,20,null,null,15,7]
    print ("The max Depth:",root)
